### PR TITLE
Tmadison/direct messaging

### PR DESCRIPTION
**Pull request by torgie-HBO**
### Change Summary
Enables direct messaging to Jenkins users

### Change Detail
Opens up a user configuration parameter called "slackUsername" that must be configured in order to successfully send direct messages. It also adds a dropdown to the job properties to specify how you would like direct messaging to behave: "Channel Only" (default), "Channel and Me", and "Me Only"

waterfall

### CODE REVIEW

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 1]
```diff
@@ -46,22 +36,24 @@ public ActiveNotifier(SlackNotifier notifier, BuildListener listener) {
 
     private SlackService getSlack(AbstractBuild r) {
         AbstractProject<?, ?> project = r.getProject();
-        String projectRoom = Util.fixEmpty(project.getProperty(SlackNotifier.SlackJobProperty.class).getRoom());
-        String teamDomain = Util.fixEmpty(project.getProperty(SlackNotifier.SlackJobProperty.class).getTeamDomain());
-        String token = Util.fixEmpty(project.getProperty(SlackNotifier.SlackJobProperty.class).getToken());
+        String projectRoom = Util.fixEmpty(project.getProperty(SlackJobProperty.class).getRoom());
+        String teamDomain = Util.fixEmpty(project.getProperty(SlackJobProperty.class).getTeamDomain());
+        String token = Util.fixEmpty(project.getProperty(SlackJobProperty.class).getToken());
+
```
Comment (siggimoo) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Any particular reason for extracting SlackJobProperty from its containing class?

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 2]
```diff
@@ -46,22 +36,24 @@ public ActiveNotifier(SlackNotifier notifier, BuildListener listener) {
...
+        String directMessage = Util.fixEmpty(project.getProperty(SlackJobProperty.class).getSendDirectMessage());
 
         EnvVars env = null;
         try {
             env = r.getEnvironment(listener);
         } catch (Exception e) {
             listener.getLogger().println("Error retrieving environment vars: " + e.getMessage());
-            env = new EnvVars();
         }
         teamDomain = env.expand(teamDomain);
```
Comment (siggimoo) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java:48]: By removing line 58, line 60 will throw a NullPointerException if there's any problem retrieving the environment.

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 3]
```diff
@@ -138,7 +130,7 @@ public void completed(AbstractBuild r) {
 
     String getChanges(AbstractBuild r) {
         if (!r.hasChangeSetComputed()) {
-            logger.info("No change set computed...");
+            logger.log(Level.FINER, "No change set computed...");
```
Comment (siggimoo) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: logger.finer() should do the same thing in a more readable fashion.

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 2 above)
Comment (torgie-HBO) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Ah, I meant to move the instantiation to line 45 to replace `EnvVars env = null;` but I must have been moving too fast.

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 3 above)
Comment (torgie-HBO) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: All logger calls have been converted unless they needed to use .log()'s Throwable parameter

**Comment by torgie-HBO** [comment]
This has been patched into a cleaner branch without so much refactoring history

**Pull request closed without merging**